core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/IdempotentAsyncResponseHandler.java

```diff
@@ -0,0 +1,117 @@
+ *
+ * One application of this wrapper is to ensure that prepare() is not called on the underlying response handler more
+ * than once per retry no matter where or how many times its invoked.
+ *
+ * This class is asserted to be thread-safe and it should be fine to have multiple threads call prepare()
+ * simultaneously.
+ *
+ * @param <T> The type of the wrapped {@link TransformingAsyncResponseHandler}
+ * @param <R> The type of the object used to determine scope.
+ */
+@SdkInternalApi
+@ThreadSafe
+public class IdempotentAsyncResponseHandler<T, R> implements TransformingAsyncResponseHandler<T> {
+    private final Supplier<R> newScopeSupplier;
+    private final BiPredicate<R, R> newScopeInRangePredicate;
+    private final TransformingAsyncResponseHandler<T> wrappedHandler;
+
+    // Taking advantage of the volatile properties of AtomicReference
+    private final AtomicReference<R> cachedScope = new AtomicReference<>();
+    private final AtomicReference<CompletableFuture<T>> cachedPreparedFuture = new AtomicReference<>();
+
+    private IdempotentAsyncResponseHandler(TransformingAsyncResponseHandler<T> wrappedHandler,
+                                           Supplier<R> newScopeSupplier,
+                                           BiPredicate<R, R> newScopeInRangePredicate) {
+        this.newScopeSupplier = newScopeSupplier;
+        this.newScopeInRangePredicate = newScopeInRangePredicate;
+        this.wrappedHandler = wrappedHandler;
+    }
+
+    /**
+     * Creates a new wrapped {@link TransformingAsyncResponseHandler}
+     * @param wrappedHandler A different handler to wrap.
+     * @param preparedScopeSupplier A function that retrieves the current scope to determine if a new future can be
+     *                              prepared.
+     * @param scopeInRangePredicate A predicate that can be used to test the cached scope against the newly retrieved
+     *                             scope to determine if a new future can be prepared.
+     * @param <T> The type of the wrapped {@link TransformingAsyncResponseHandler}
+     * @param <R> The type of the object used to determine scope.
+     */
+    public static <T, R> IdempotentAsyncResponseHandler<T, R> create(
+            TransformingAsyncResponseHandler<T> wrappedHandler,
+            Supplier<R> preparedScopeSupplier,
+            BiPredicate<R, R> scopeInRangePredicate) {
+
+        return new IdempotentAsyncResponseHandler<>(wrappedHandler,
+                                                    preparedScopeSupplier,
+                                                    scopeInRangePredicate);
+    }
+
+    @Override
+    public CompletableFuture<T> prepare() {
+        // Test if the new scope is in range of the cached scope; if not delegate and cache result and new scope
+        if (cachedPreparedFuture.get() == null ||
+            !newScopeInRangePredicate.test(newScopeSupplier.get(), cachedScope.get())) {
+
+            synchronized (this) {
```

Comment:
I was having the same conversation with Sam. The thread-safety was force of habit to increase the re-usability of the class. It's unclear how much penalty we're actually paying for it since prepare() only gets called twice for most requests. I'm keeping it in our back pocket if we need to trim any millis on the perf tests once we're done with everything else.

Reply:
My concern it that this is going to affect any async client. We can run some benchmarks to get an idea of the penalty and it'd be nice to understand how much it's going to affect the performance w/ the change. 

Yeah, agreed it's not a blocker for the PR, but let's get the benchmark numbers diff and create a backlog item to track it.